feat: add new adaptive termination type

Christopher-Chianelli opened the pull request:
The adaptive termination type has the concept of a grace period G and minimum improvement ratio R.

The adaptive termination type works as follows.

- When in the grace period, do not terminate.

- If a non-lowest score level changes at any point, reset the grace period (or enter a new one if the grace period is finished).

- After the duration G of the grace period passes, record the difference of final best score and the starting best score as Y_0. If this is 0, terminate.

- On each step, calculate the difference between the current best score and the best score G moments ago as Y_n. If Y_n / Y < R (i.e. the score did not improve by at least R% of the improvement in the grace window), terminate.

Fixes #1311

Christopher-Chianelli committed:
feat: add new adaptive termination type

The adaptive termination type has the concept of a grace period G
and minimum improvement ratio R.

The adaptive termination type works as follows.

- When in the grace period, do not terminate.

- If a non-lowest score level changes at any point, reset the grace period
  (or enter a new one if the grace period is finished).

- After the duration G of the grace period passes, record the
  difference of final best score and the starting best score as
  Y_0. If this is 0, terminate.

- On each step, calculate the difference between the current best score
  and the best score G moments ago as Y_n. If Y_n / Y < R
  (i.e. the score did not improve by at least R% of the improvement
   in the grace window), terminate.

Christopher-Chianelli committed:
fix: set adaptive config to null and include it in the isConfigured check

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java:
```diff
@@ -0,0 +1,171 @@
+    public AdaptiveTermination(long gracePeriodMillis, double minimumImprovementRatio) {
+        this.gracePeriodMillis = gracePeriodMillis;
+        this.minimumImprovementRatio = minimumImprovementRatio;
+        this.scoresByTime = new TreeMap<>();
```
I was skeptical of this approach, as fast-stepping algorithms will easily put tens of thousands of entries to this map per second, and I was expecting that to be both CPU-intensive and memory-intensive.

However, in the experiments I ran, this did not materialize. Impact of the termination is barely visible in the data. Interesting.

triceo replied:
I wonder though if the `TreeMap` is bringing any tangible benefits. Consider this:

- We don't actually need sorting, do we? The values are sequential, a `List` would do.
- We are forced to box `long` into `Long`, causing some unnecessary GC pressure.

I wonder if we could convert this into `List<Score_>`. The index would have to become relative to the last reset, so starting with zero.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java:
```diff
@@ -0,0 +1,171 @@
+        this.scoresByTime = new TreeMap<>();
+    }
+
+    private record LevelScoreDiff(boolean harderScoreChanged, double softestScoreDiff) {
```
Typically, I would be all for a carrier type like this. In this case, and on the hot path, I'd rather store the two variables individually and avoid allocating the instance of the record. 

In fact, this is the only thing that actually made a measurable impact in my experiments - `LevelScoreDiff` was ~10th most allocated object on the heap.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java:
```diff
@@ -0,0 +1,171 @@
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean isSolverTerminated(SolverScope<Solution_> solverScope) {
+        return isTerminated(System.currentTimeMillis(), (Score_) solverScope.getBestScore());
```
I wonder if we should use nanotime instead, and truncate it to millis.
The precision of `currentTimeMillis` is questionable, see Javadoc:

> Note that while the unit of time of the return value is a millisecond, the granularity of the value depends on the underlying operating system and may be larger. For example, many operating systems measure time in units of tens of milliseconds.

triceo replied:
Also... should this terminate the solver even in, for example, a CH phase? Arguably it should not, the CH phase generally doesn't improve.

The way how unimproved terminations get around it is by forcing the adaptive termination to be only possible in local search phase. Maybe we do something like that too?

Although I really don't like that, because it causes config pain. (You need to define LS phase, and put the termination inside of it.) Maybe we refactor the terminations a bit, so that it is possible for an unimproved termination to be defined on the solver-level, but never terminate a CH? Would need a discussion, I think.

(My ultimate goal is to have only one termination. This one. And this configuration quirk is preventing that from happening.)

Christopher-Chianelli replied:
It cannot terminate in the CH phase, because the CH ALWAYS improve (namely, it improves the init score every step, which resets the grace period).

triceo replied:
Not necessarily true - list var with allows unassigned has init score always zero.
But doesn't matter - this termination, for now, is only allowed on the LS phase.

Christopher-Chianelli replied:
CH in unassigned does not decrease the best score; it only can increase it (it can leave the value unassigned (no increase), or assign a value (which more than likely will increase the medium score).

triceo replied:
Those are assumptions - as long as the thing is technically possible, we need to have an answer. In this case, this (along with the unimproved terminations) will be exempt from terminating the CH, once it's even possible to configure them on the top level. For now, until we implement that, it's not a problem at all.

In fact, I am surprised why this termination even cares about init score. How does it receive events from CH, if it needs to be configured on the LS only? That smells like a bug.

Christopher-Chianelli replied:
It was from `solvingStarted` (where the best score is typically uninitialized). All terminations gets the `solvingStarted` event even if configured on a phase level.

Christopher-Chianelli replied:
(in short, the termination was incorrectly using `solvingStarted` instead of `phaseStarted` to initialize).

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java:
```diff
@@ -0,0 +1,171 @@
+
+    @Override
+    public double calculateSolverTimeGradient(SolverScope<Solution_> solverScope) {
+        return -1.0;
```
I wonder if there's a way to estimate this based on how close we are to the minimum necessary improvement.

triceo reviewed (commented):
That was quick! Leaving suggestions for improvement.

Maybe we introduce some TRACE-level logging? This termination is doing smart things, and some visibility into the mechanism may be necessary. We should find a decent way of logging the decrease of the ratio - maybe log every time in halves?

Also required:
- Documentation. Preferrably with a visualization as above.
- Naming. (Already started a conversation on Slack.)
- Add this to quickstarts?

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java:
```diff
@@ -0,0 +1,171 @@
+import ai.timefold.solver.core.impl.solver.scope.SolverScope;
+import ai.timefold.solver.core.impl.solver.thread.ChildThreadType;
+
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
```
You can use `@NullMarked` which means that everything is `@NonNull` unless marked as `@Nullable`. In my experience, this saves a lot of hassle.

Christopher-Chianelli committed:
chore: Use ring buffers and signals to improve performance of adaptive

- Use a signal exception to avoid the need for a carrier type, allowing
  use to return double instead of Double or LevelScoreDiff

- Use a pair of ring buffers to represent the TreeMap<> to avoid
  creating boxed types and to optimize on the fact the entries we
  are looking for are often near the start, and we are always
  appending to the end

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java:
```diff
@@ -1,57 +1,71 @@
+        var softestLevel = scoreDiffs.length - 1;
+        for (int i = 0; i < softestLevel; i++) {
+            if (scoreDiffs[i] != 0.0) {
+                throw new HardLevelImprovedSignal();
```
May I suggest we return an impossible value instead of using exceptions for control flow? Perhaps [NaN](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Double.html#NaN)?

triceo replied:
I understand your reasoning, but I disagree. In my opinion, what you're doing here is misusing exceptions - you want to terminate without a return, and exceptions are a _nice_ way of doing it. When you have a hammer, everything becomes a nail.

Exceptions are meant for _exceptional_ cases - this situation is not exceptional.
Exceptions are meant for reporting errors, not for control flow. This is not an error.
And finally, unless I'm mistaken, exceptions mess with JVM's optimizations, because they abort the flow.

I could go on, but I'd just be duplicating what was already said much better in [the Java bible](https://www.amazon.com/Effective-Java-Joshua-Bloch/dp/0134685997). :-)

triceo replied:
Good, data is always better than no data. Doesn't this in fact prove that using the exception is the slower option, as I expected?

> that there is no way to enforce the caller to actually check it is NaN.

The argument would have value if this was public API where there was a risk that the user can misuse it. However, this is a private method inside of an internal implementation - external users don't enter the picture, and for our purposes we make sure it works through tests.

Christopher-Chianelli replied:
> Good, data is always better than no data. Doesn't this in fact prove that using the exception is the slower option, as I expected?

Yes (although I wouldn't use prove; this is only a single benchmark run, on a single machine).

> > that there is no way to enforce the caller to actually check it is NaN.
> 
> The argument would have value if this was public API where there was a risk that the user can misuse it. However, this is a private method inside of an internal implementation - external users don't enter the picture, and for our own purposes we make sure it works through tests.

I will rename the method to something like `softImprovementOrNaNOnHardImprovement` to prevent headaches for future me.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java:
```diff
@@ -1,57 +1,71 @@
+
+    private static <Score_ extends Score<Score_>> double softImprovement(@NonNull Score_ start,
+            @NonNull Score_ end) throws HardLevelImprovedSignal {
+        if (start.equals(end)) {
```
Can the score be _worse_ here? Arguably, it can - sometimes the solver picks a worse move to try to get out of a local optima.

Christopher-Chianelli replied:
No, it uses best score, not step score. Best score should never decrease.
```
stepScope.getPhaseScope().getBestScore()
```

Christopher-Chianelli committed:
chore: fix sonar issues, add more tests

Christopher-Chianelli committed:
chore: fix Sonar issues

Christopher-Chianelli committed:
chore: use isZero instead of isEqualTo(0)

Christopher-Chianelli committed:
chore: Return NaN instead of throwing an exception on harder level change

Christopher-Chianelli committed:
fix: add new AdaptiveConfig class to reflect-config.json

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveScoreRingBuffer.java:
```diff
@@ -0,0 +1,203 @@
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
+
+final class AdaptiveScoreRingBuffer<Score_ extends Score<Score_>> {
```
Is there a way to refactor the ring buffer to make it a general-purpose collection?
I can see it being useful to the Solver for more than this.

Christopher-Chianelli replied:
Yes, but then we need to box the longs (unless you want to keep it as a map with long keys?) . Additionally it not a typical ring buffer (which normally acts a queue); its method removes all elements up to the floored query (and keeps the query entry). It also relies on the fact the queries will always increase due to the behaviour of `System.nanoTime`.

If you can spec out what you want the RingBuffer to be able to do, I can refactor it to match.

triceo replied:
Nevermind then - the generalization would hurt the primary purpose.
But the description of differences from a "typical" ring buffer would make for a nice Javadoc.

Christopher-Chianelli committed:
fix: reset grace when init score changes

Christopher-Chianelli committed:
chore: rename to diminished returns termination + add quarkus property

Note: it is impossible to test a startup failure due to a runtime
property in Quarkus; see https://github.com/quarkusio/quarkus/issues/45669
for details.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/DiminishedReturnsScoreRingBuffer.java:
```diff
@@ -0,0 +1,224 @@
+        scoreRingBuffer = newScoreRingBuffer;
+    }
+
+    public void clear() {
```
I wonder if:

- Throwing away the array and re-allocating it would be faster than clearing it. (Maybe not.)
- We should allocate the array using the original size, not using whatever size it is now.

Christopher-Chianelli replied:
We should keep the current size; the size of this buffer is based on score calculation speed (the higher the speed, the more entries in the buffer). Since score calculation speed stays roughly the same in LS, the buffer has an ideal size that it will always regrow to if was shrunk.

Creating a new array each time creates additional memory pressure (although clearing should be rare in most cases) and may invalidate the CPU cache.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/DiminishedReturnsScoreRingBuffer.java:
```diff
@@ -0,0 +1,224 @@
+     *
+     * @return the first element of the score ring buffer
+     */
+    public @NonNull Score_ peekFirst() {
```
Since the array starts as empty, this should not be annotated as `@NonNull`, because `null` is entirely possible.

Christopher-Chianelli replied:
The buffer starts with the score at the start of a phase, and polling will always keep one element in the buffer. Thus, it is an impossible state to have the buffer empty when it's called.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/DiminishedReturnsScoreRingBuffer.java:
```diff
@@ -0,0 +1,224 @@
+     */
+    public @NonNull Score_ pollLatestScoreBeforeTimeAndClearPrior(long nanoTime) {
+        if (readIndex == writeIndex && nanoTimeRingBuffer[writeIndex] == 0L) {
+            throw new IllegalStateException("Ring buffer is empty");
```
Do you want to treat this as an exception? I see two options here:

- Either this is an exceptional state which must not happen, because the algorithm would not be correct. Then the message should start with "Impossible state: ", and I'm going to argue it should be thrown by the code calling this method, not by this method directly.
- Or this is a totally possible state, which we just want to signal to the calling code. In that case, this exception should be replaced by a return on `null` - see our previous conversation on exceptions and control flow.

Christopher-Chianelli replied:
Exceptional state which must not happen; correctness assumes there at least one element in the buffer (which is inserted on start & whenever grace is reset).

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/DiminishedReturnsTermination.java:
```diff
@@ -0,0 +1,179 @@
+
+public final class DiminishedReturnsTermination<Solution_, Score_ extends Score<Score_>>
+        extends AbstractTermination<Solution_> {
+    static final long NANOS_PER_MILLISECOND = 1_000_000;
+
+    private final long slidingWindowNanos;
```
If the length of the window was a `Duration` (see my comment on the termination config), then you could avoid this and just call `slidingWindowDuration.toMillis()`.

Christopher-Chianelli replied:
It was done like this to match how other terminations do it.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/DiminishedReturnsTermination.java:
```diff
@@ -0,0 +1,179 @@
+                    // Should be impossible; the only cases where the best score improves
+                    // and have a lower softest level are if either a harder level or init score
+                    // improves, but if that happens, we reset the grace period.
+                    throw new IllegalStateException("The score deteriorated from (%s) to (%s) during the grace period."
```
```suggestion
                    throw new IllegalStateException("Impossible state: The score deteriorated from (%s) to (%s) during the grace period."
```

triceo replied:
This is a convention we generally honor in other parts of the codebase.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/TerminationFactory.java:
```diff
@@ -122,6 +122,19 @@ The termination with bestScoreFeasible (%s) can only be used with a score type \
+            }
+            var minimumImprovementRatio = adaptiveTerminationConfig.getMinimumImprovementRatio();
+            if (minimumImprovementRatio == null) {
+                minimumImprovementRatio = 0.01;
```
See Slack thread on the "correct" default value.

triceo commented on quarkus-integration/quarkus/deployment/src/test/java/ai/timefold/solver/quarkus/TimefoldProcessorOverridePropertiesAtRuntimeTest.java:
```diff
@@ -48,6 +48,9 @@ private static String getRequiredProperty(String name) {
             .overrideConfigKey("quarkus.timefold.solver.termination.best-score-limit", "0")
             .overrideConfigKey("quarkus.timefold.solver.move-thread-count", "4")
             .overrideConfigKey("quarkus.timefold.solver-manager.parallel-solver-count", "1")
+            .overrideConfigKey("quarkus.timefold.solver.termination.diminished-returns.enabled", "false")
```
Arguably this is not necessary, if either of the other two properties is present?

Christopher-Chianelli replied:
The other two properties are optional, which make it hard to use the defaults unless there a third property.
Additionally, if you want to use diminishing returns by default, but have the option to disable it, you need to somehow delete the defined properties if there is no enabled property.

triceo replied:
I see the reason for the property existing - I'm just arguing you don't need to specify it, if you specify any of the others.

Does it get confusing as you switch from one to the other? Probably. But how often does that actually happen?

triceo commented on quarkus-integration/quarkus/deployment/src/test/java/ai/timefold/solver/quarkus/TimefoldProcessorXMLDiminishedReturnsTest.java:
```diff
@@ -0,0 +1,24 @@
+package ai.timefold.solver.quarkus;
+
+class TimefoldProcessorXMLDiminishedReturnsTest {
```
Delete or finish?

triceo commented on quarkus-integration/quarkus/runtime/src/main/java/ai/timefold/solver/quarkus/config/DiminishedReturnsRuntimeConfig.java:
```diff
@@ -0,0 +1,44 @@
+     * <br/>
+     * Defaults to 0.1.
+     */
+    Optional<Double> minimumImprovementRatio();
```
`OptionalDouble` exists.
(For some reason, `OptionalBoolean` does not.)

Christopher-Chianelli replied:
As for why `OptionalBoolean` does not exist: https://mail.openjdk.org/pipermail/lambda-dev/2012-November/006750.html

triceo commented on quarkus-integration/quarkus/runtime/src/main/java/ai/timefold/solver/quarkus/config/DiminishedReturnsRuntimeConfig.java:
```diff
@@ -0,0 +1,44 @@
+     * the current best score and the best score from 30 seconds ago is less than
+     * 25soft (= 0.25 * 100soft).
+     * <br/>
+     * Defaults to 0.1.
```
Please remember to update.

triceo reviewed (commented):
Looking good!

Still need to add documentation. Since this termination is non-trivial, we'll need an illustration (akin to what I provided above) together with an explanation of the concepts, and broad strokes of the algorithm.

Christopher-Chianelli commented on core/src/main/java/ai/timefold/solver/core/config/solver/termination/DiminishedReturnsTerminationConfig.java:
```diff
@@ -0,0 +1,210 @@
+                || slidingWindowDays != null;
+    }
+
+    public @Nullable Long calculateSlidingWindowMilliseconds() {
+        if (slidingWindowMilliseconds == null && slidingWindowSeconds == null
+                && slidingWindowMinutes == null && slidingWindowHours == null
+                && slidingWindowDays == null) {
+            if (slidingWindowDuration != null) {
+                if (slidingWindowDuration.getNano() % 1000 != 0) {
+                    throw new IllegalArgumentException("The termination slidingWindowDuration (" + slidingWindowDuration
+                            + ") cannot use nanoseconds.");
+                }
+                return slidingWindowDuration.toMillis();
+            }
+            return null;
+        }
+        if (slidingWindowDuration != null) {
+            throw new IllegalArgumentException("The termination slidingWindowDuration (" + slidingWindowDuration
+                    + ") cannot be combined with slidingWindowMilliseconds (" + slidingWindowMilliseconds
+                    + "), slidingWindowSeconds (" + slidingWindowSeconds
+                    + "), slidingWindowMinutes (" + slidingWindowMinutes
+                    + "), slidingWindowHours (" + slidingWindowHours + "),"
+                    + ") or slidingWindowDays (" + slidingWindowDays + ").");
+        }
+        long slidingWindowMillis = 0L
+                + requireNonNegative(slidingWindowMilliseconds, "slidingWindowMilliseconds")
+                + requireNonNegative(slidingWindowSeconds, "slidingWindowSeconds") * 1000L
+                + requireNonNegative(slidingWindowMinutes, "slidingWindowMinutes") * 60_000L
+                + requireNonNegative(slidingWindowHours, "slidingWindowHours") * 3_600_000L
+                + requireNonNegative(slidingWindowDays, "slidingWindowDays") * 86_400_000L;
+        return slidingWindowMillis;
+    }
```
This is how the other terminations do this.

Christopher-Chianelli commented on core/src/main/java/ai/timefold/solver/core/impl/solver/termination/DiminishedReturnsScoreRingBuffer.java:
```diff
@@ -0,0 +1,224 @@
+ * @param <Score_> The score type
+ */
+final class DiminishedReturnsScoreRingBuffer<Score_ extends Score<Score_>> {
+    private static final int DEFAULT_CAPACITY = 4096;
```
Arbitrary but commonly used; 4096 is the page size in the linux kernel, the typical size of a file system sector, etc. It is also a common buffer size used in Java libraries. Want to use a power of two for alignment + probably faster modulo.

Christopher-Chianelli committed:
chore: review suggestions, update default ratio, add tests for quarkus property

Christopher-Chianelli committed:
chore: add diminished-returns property support to Spring

Christopher-Chianelli committed:
docs: add docs for diminished returns

triceo commented on docs/src/modules/ROOT/pages/optimization-algorithms/overview.adoc:
```diff
@@ -637,6 +637,126 @@ This is useful for benchmarking.
+==== `DiminishedReturnsTermination`
+
+Terminates when the rate of improvement is below a percentage of the initial rate of improvement.
+The rate of improvement is the score difference between the final best score and the initial best score in a window, divided by the duration of the window.
```
I think we should talk about the softest score here.

triceo commented on docs/src/modules/ROOT/pages/optimization-algorithms/overview.adoc:
```diff
@@ -637,6 +637,126 @@ This is useful for benchmarking.
+
+image::optimization-algorithms/overview/diminishedReturnsTermination.png[align="center"]
+
+[NOTE]
+====
+If the hard score improves, the grace period is reset and the initial rate of improvement is recalculated.
+====
```
Since this is a core part of the algorithm, I don't think it should be a NOTE. Rather, a normal text like everything else.

Also, the above comment about the softest score will pay off here, where we talk about the hard score all of a sudden.

triceo commented on docs/src/modules/ROOT/pages/optimization-algorithms/overview.adoc:
```diff
@@ -637,6 +637,126 @@ This is useful for benchmarking.
+  </localSearch>
+----
++
+** Seconds
++
+[source,xml,options="nowrap"]
+----
+  <localSearch>
+    <termination>
+      <diminishedReturns>
+        <slidingWindowSeconds>500</slidingWindowSeconds>
+      </diminishedReturns>
+    </termination>
+  </localSearch>
+----
```
I consider this section too verbose. It's IMO enough to show the duration and the seconds, and then mention we also support the others.

Christopher-Chianelli replied:
Copied from unimproved time spent termination, which list them all: https://docs.timefold.ai/timefold-solver/latest/optimization-algorithms/overview#unimprovedTimeMillisSpentTermination

triceo commented on docs/src/modules/ROOT/pages/optimization-algorithms/overview.adoc:
```diff
@@ -637,6 +637,126 @@ This is useful for benchmarking.
+  </localSearch>
+----
+
+This `Termination` can only be used for a `Phase` (such as ``<localSearch>``), not for the `Solver` itself.
```
I'd invert this statement - this termination cannot be used globally on the solver, only on a phase, and not on a CH phase.

Christopher-Chianelli replied:
Copied from step count termination https://docs.timefold.ai/timefold-solver/latest/optimization-algorithms/overview#stepCountTermination

Christopher-Chianelli replied:
And it can be used on a CH phase (no terminations are exclusive to LS; all phase exclusive terminations will work on any phase). We are also inconsistent in how the unimproved terminations handle it; unimproved time spent is allowed to be configured on a solver level (and just ignores the CH: https://github.com/TimefoldAI/timefold-solver/blob/4b446075bea81eea393edd887a32960470e2970b/core/src/main/java/ai/timefold/solver/core/impl/solver/termination/UnimprovedTimeMillisSpentScoreDifferenceThresholdTermination.java#L61-L71), whereas unimproved step count must be configured on phases (https://github.com/TimefoldAI/timefold-solver/blob/4b446075bea81eea393edd887a32960470e2970b/core/src/main/java/ai/timefold/solver/core/impl/solver/termination/UnimprovedStepCountTermination.java#L25-L26).

triceo replied:
This is interesting - the inconsistency is surprising me. I thought the fail-fast was present in all the unimproved terminations. We'll get to fix that eventually.

triceo commented on docs/src/modules/ROOT/pages/optimization-algorithms/overview.adoc:
```diff
@@ -637,6 +637,126 @@ This is useful for benchmarking.
+
+This `Termination` can only be used for a `Phase` (such as ``<localSearch>``), not for the `Solver` itself.
 
 [#combiningMultipleTerminations]
```
I'd mention that the termination can run for a very long time, and add a recommendation to combine the termination with a maximum time spent, to ensure it doesn't run for longer than acceptable.

Do we have a test for this? When I was toying with the termination, I saw some weird behavior where a composite termination wasn't working properly.

Christopher-Chianelli replied:
[We have AND/OR composition tests](https://github.com/TimefoldAI/timefold-solver/blob/8ca431c4a6a2edf4b74f903d8aa40a45f5b0355d/core/src/test/java/ai/timefold/solver/core/impl/solver/termination/AndCompositeTerminationTest.java#L15). When I tried composing it with a time spent termination, it terminated as expected (did one with a short time spent, long grace, and another with a long time spent, short grace; both terminated quickly).

triceo commented on docs/src/modules/ROOT/pages/optimization-algorithms/overview.adoc:
```diff
@@ -637,6 +637,126 @@ This is useful for benchmarking.
 Switching xref:using-timefold-solver/running-the-solver.adoc#environmentMode[EnvironmentMode] can heavily impact when this termination ends.
 
+[#diminishedReturnsTermination]
+==== `DiminishedReturnsTermination`
```
I'd start with an introductory paragraph explaining why we've done this. Absolute terminations are hard to configure, because they either depend on hardware (time-based) or on the problem (score-based). The relative termination doesn't have this problem.

triceo reviewed (approved):
LGTM after comments resolved.
Looking forward to using this one!

Christopher-Chianelli committed:
chore: add diminished returns properties to config-properties, review suggestions
